### PR TITLE
Updating the rewarded adapter to use the new Rewarded API

### PR DESCRIPTION
**Pull request by joshuajplagonera**
Also updated adapters to allow publishers to pass in flags for Child Directed Treatment and Under Age of Consent flags.

### CODE REVIEW

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java [hunk 1]
```diff
@@ -105,6 +107,29 @@ protected void loadBanner(
         // Google's personalization preference. Publishers should work with Google to be GDPR-compliant.
         forwardNpaIfSet(builder);
 
+        // Publishers may want to indicate that their content is child-directed and forward this
+        // information to Google.
+        if (localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY) != null) {
```
Comment (ericleich) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: localExtras is <String, Object>. Could publishers pass in a boolean instead of a string that we convert to boolean?

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 2]
```diff
@@ -133,31 +116,14 @@ protected boolean checkAndInitializeSdk(@NonNull Activity launcherActivity,
...
             throws Exception {
         if (!sIsInitialized.getAndSet(true)) {
-
             if (TextUtils.isEmpty(serverExtras.get(KEY_EXTRA_APPLICATION_ID))) {
                 MobileAds.initialize(launcherActivity);
             } else {
                 MobileAds.initialize(launcherActivity, serverExtras.get(KEY_EXTRA_APPLICATION_ID));
             }
 
-            if (TextUtils.isEmpty(serverExtras.get(KEY_EXTRA_AD_UNIT_ID))) {
```
Comment (ericleich) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Don't we still want this? If ad unit is missing, that's a problem.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 3]
```diff
@@ -29,6 +29,8 @@
     private static final String ADAPTER_NAME = GooglePlayServicesInterstitial.class.getSimpleName();
     private static final String CONTENT_URL_KEY = "contentUrl";
     private static final String TEST_DEVICES_KEY = "testDevices";
+    private static final String TAG_FOR_CHILD_DIRECTED_KEY = "tagForChildDirectedTreatment";
+    private static final String TAG_FOR_UNDER_AGE_OF_CONSENT_KEY = "tagForUnderAgeOfConsent";
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Let's organize these constants alphabetically.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 4]
```diff
@@ -94,6 +96,29 @@ protected void loadInterstitial(
         // Google's personalization preference. Publishers should work with Google to be GDPR-compliant.
         forwardNpaIfSet(builder);
 
+        // Publishers may want to indicate that their content is child-directed and forward this
+        // information to Google.
+        if (localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY) != null) {
+            String childDirected = localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY).toString();
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Comment from the banner class applies here regarding the NPE.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 5]
```diff
@@ -94,6 +96,29 @@ protected void loadInterstitial(
...
+        }
+
+        // Publishers may want to mark your their requests to receive treatment for users in the
+        // European Economic Area (EEA) under the age of consent.
+        if (localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY) != null) {
+            String underAgeOfConsent = localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY).toString();
+            if (!TextUtils.isEmpty(underAgeOfConsent)) {
+                boolean flag = Boolean.parseBoolean(underAgeOfConsent);
+                if (flag) {
+                    builder.setTagForUnderAgeOfConsent(AdRequest.TAG_FOR_UNDER_AGE_OF_CONSENT_TRUE);
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: The constant can be statically imported.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 6]
```diff
@@ -31,13 +33,13 @@
...
-public class GooglePlayServicesRewardedVideo extends CustomEventRewardedVideo implements
-        RewardedVideoAdListener {
+public class GooglePlayServicesRewardedVideo extends CustomEventRewardedVideo {
 
     /**
      * String to represent the simple class name to be used in log entries.
      */
-    private static final String ADAPTER_NAME = GooglePlayServicesRewardedVideo.class.getSimpleName();
+    private static final String ADAPTER_NAME =
+            GooglePlayServicesRewardedVideo.class.getSimpleName();
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: This can be moved back to be on the same line since it doesn't go past the hard wrap guide in Android Studio.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 7]
```diff
@@ -57,13 +59,17 @@
     /**
      * Google Mobile Ads rewarded video ad unit ID.
      */
-    @NonNull
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Why is this removed?

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 2 above)
Comment (chauduyphanvu1) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Let's add back the check for empty/null ad unit ID, and bail out (log + return) if necessary.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 8]
```diff
@@ -177,7 +143,8 @@ protected void loadWithSdkInitialized(@NonNull final Activity activity,
             public void run() {
                 isAdLoaded = false;
 
-                if (TextUtils.isEmpty(serverExtras.get(KEY_EXTRA_AD_UNIT_ID))) {
+                mAdUnitId = serverExtras.get(KEY_EXTRA_AD_UNIT_ID);
+                if (TextUtils.isEmpty(mAdUnitId)) {
                     // Using class name as the network ID for this callback since the ad unit ID is
                     // invalid.
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: `Using class name as the network ID for this callback since the ad unit ID is invalid` is not applicable - it's OK to pass an empty `mAdUnitId` to this failure callback.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 9]
```diff
@@ -186,67 +153,81 @@ public void run() {
...
+                // Publishers should work with Google to be GDPR-compliant.
+                forwardNpaIfSet(builder);
 
-                    AdRequest adRequest = builder.build();
-                    mRewardedVideoAd.loadAd(mAdUnitId, adRequest);
+                // Publishers may want to indicate that their content is child-directed and
+                // forward this information to Google.
+                Boolean isTFCD = GooglePlayServicesMediationSettings.isTaggedForChildDirectedTreatment();
+                if (isTFCD != null) {
+                    builder.tagForChildDirectedTreatment(isTFCD);
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: `tagForChildDirectedTreatment` accepts a `boolean`, not `Boolean`.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 10]
```diff
@@ -186,67 +153,81 @@ public void run() {
...
-                mAdUnitId = serverExtras.get(KEY_EXTRA_AD_UNIT_ID);
 
-                if (mRewardedVideoAd == null) {
-                    mRewardedVideoAd = MobileAds.getRewardedVideoAdInstance(activity);
-                    mRewardedVideoAd.setRewardedVideoAdListener(GooglePlayServicesRewardedVideo.this);
-                }
+                mWeakActivity = new WeakReference<>(activity);
+                mRewardedAd = new RewardedAd(activity, mAdUnitId);
 
-                if (mRewardedVideoAd.isLoaded()) {
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Since we're not calling `isLoaded()` here, we can remove the `new Handler(Looper.getMainLooper()).post(new Runnable()` since that was added to force calling `isLoaded()` on the main thread to avoid crashes.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 11]
```diff
@@ -186,67 +153,81 @@ public void run() {
...
-            builder.addNetworkExtrasBundle(AdMobAdapter.class, GooglePlayServicesMediationSettings.getNpaBundle());
+            builder.addNetworkExtrasBundle(AdMobAdapter.class,
+                    GooglePlayServicesMediationSettings.getNpaBundle());
         }
     }
 
     @Override
     protected boolean hasVideoAvailable() {
-        return mRewardedVideoAd != null && isAdLoaded;
+        return mRewardedAd != null && isAdLoaded;
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: We might want to use `return mRewardedAd != null && mRewardedAdisLoaded();`

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 12]
```diff
@@ -377,6 +387,16 @@ public GooglePlayServicesMediationSettings(Bundle bundle, String url, String id)
             testDeviceId = id;
         }
 
+        public GooglePlayServicesMediationSettings(Bundle bundle,
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Let's clean this up and have publishers pass everything into one `Bundle` (instead of separately like a `String` or a `Boolean`). We can then delete the other constructors, and just parse for the right data from the same `Bundle`.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 2 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Change reverted.

File: AdMob/CHANGELOG.md [hunk 13]
```diff
@@ -1,4 +1,7 @@
 ## Changelog
+  * 17.2.0.0
+    * Update the rewarded ad adapter to use Google's new Rewarded Video API.
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/CHANGELOG.md]: Added.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 8 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Removed comment.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 10 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Removed Handler.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 11 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Added.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java (hunk 1 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: Updated to parse as boolean

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 9 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Fixed to use .booleanValue()

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 12 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Updated constructor.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 14]
```diff
@@ -354,27 +355,65 @@ private MoPubErrorCode getMoPubErrorCode(int error) {
...
+                testDeviceId = bundle.getString(TEST_DEVICES_KEY);
+            }
 
-        public GooglePlayServicesMediationSettings(Bundle bundle, String url, String id) {
-            npaBundle = bundle;
-            contentUrl = url;
-            testDeviceId = id;
+            if (bundle.containsKey(NPA_BUNDLE_KEY)) {
+                npaBundle = bundle.getBundle(NPA_BUNDLE_KEY);
+            }
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: This check is not very clear. Publishers will pass in a bundle with any number of entries, and `npa:n` might exist. If the adapter passes the `npa` value to Google, the constructor should be extracting a String from the `Bundle` instead of calling `getBundle()`. At the time of creating an ad request builder, we can then constructor a new `Bundle` to be sent. So:

* The `npaBundle` variable should be a `String` instead of a `Bundle`. And let's rename the variable to something like `npaValue`.
* The `NPA_BUNDLE_KEY` constant should have the value of `npa` instead of `npaBundle`. And let's rename the variable to something like `NPA_KEY`.
* We should check `npaValue = bundle.getString(NPA_KEY);`. It might be null/empty, `0`, or `1`.
* In `forwardNpaIfSet()`, check if `npaValue` is null/empty. If it's not, create a `Bundle` there to be passed to `AdRequest.Builder.addNetworkExtrasBundle()`.

Doing it that way avoids having multiple `Bundle`s to deal with.

File: AdMob/CHANGELOG.md [hunk 15]
```diff
@@ -1,4 +1,10 @@
 ## Changelog
+  * 17.2.0.0
+    * Update the rewarded ad adapter to use AdMob's new Rewarded Video API
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/CHANGELOG.md]: Done.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java [hunk 16]
```diff
@@ -105,6 +110,24 @@ protected void loadBanner(
...
+            boolean childDirected = (boolean) localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: Updated.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java [hunk 17]
```diff
@@ -105,6 +110,24 @@ protected void loadBanner(
...
+            builder.tagForChildDirectedTreatment(childDirected);
+        }
+
+        // Publishers may want to mark their requests to receive treatment for users in the
+        // European Economic Area (EEA) under the age of consent.
+        if (localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY) != null) {
+            boolean underAgeOfConsent = (boolean) localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: Updated.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 18]
```diff
@@ -94,6 +99,24 @@ protected void loadInterstitial(
         // Google's personalization preference. Publishers should work with Google to be GDPR-compliant.
         forwardNpaIfSet(builder);
 
+        // Publishers may want to indicate that their content is child-directed and forward this
+        // information to Google.
+        if (localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY) != null) {
+            boolean childDirected = (boolean) localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Updated.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 19]
```diff
@@ -94,6 +99,24 @@ protected void loadInterstitial(
...
+            builder.tagForChildDirectedTreatment(childDirected);
+        }
+
+        // Publishers may want to mark their requests to receive treatment for users in the
+        // European Economic Area (EEA) under the age of consent.
+        if (localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY) != null) {
+            boolean underAgeOfConsent = (boolean) localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Updated.

File: AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java [hunk 20]
```diff
@@ -451,6 +465,24 @@ public void onAdFailedToLoad(int errorCode) {
             // Google's personalization preference. Publishers should work with Google to be GDPR-compliant.
             forwardNpaIfSet(requestBuilder);
 
+            // Publishers may want to indicate that their content is child-directed and forward this
+            // information to Google.
+            if (localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY) != null) {
+                boolean childDirected = (boolean) localExtras.get(TAG_FOR_CHILD_DIRECTED_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java]: Updated.

File: AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java [hunk 21]
```diff
@@ -451,6 +465,24 @@ public void onAdFailedToLoad(int errorCode) {
...
+                requestBuilder.tagForChildDirectedTreatment(childDirected);
+            }
+
+            // Publishers may want to mark their requests to receive treatment for users in the
+            // European Economic Area (EEA) under the age of consent.
+            if (localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY) != null) {
+                boolean underAgeOfConsent = (boolean) localExtras.get(TAG_FOR_UNDER_AGE_OF_CONSENT_KEY);
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java]: Updated.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 22]
```diff
@@ -166,87 +166,91 @@ protected boolean checkAndInitializeSdk(@NonNull Activity launcherActivity,
...
-                    // Consent collected from the MoPub’s consent dialogue should not be used to set up
-                    // Google's personalization preference. Publishers should work with Google to be GDPR-compliant.
-                    forwardNpaIfSet(builder);
-
-                    AdRequest adRequest = builder.build();
-                    mRewardedVideoAd.loadAd(mAdUnitId, adRequest);
-
-                    MoPubLog.log(getAdNetworkId(), LOAD_ATTEMPTED, ADAPTER_NAME);
-                }
+        if (TextUtils.isEmpty(serverExtras.get(KEY_EXTRA_AD_UNIT_ID))) {
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Used `mAdUnitId`. However, removed its `@Nonnull` property.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 23]
```diff
@@ -166,87 +166,91 @@ protected boolean checkAndInitializeSdk(@NonNull Activity launcherActivity,
...
+            MoPubRewardedVideoManager.onRewardedVideoLoadFailure(
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Added

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 24]
```diff
@@ -166,87 +166,91 @@ protected boolean checkAndInitializeSdk(@NonNull Activity launcherActivity,
...
+        // Consent collected from the MoPub’s consent dialogue should not be used
+        // to set up Google's personalization preference.
+        // Publishers should work with Google to be GDPR-compliant.
+        forwardNpaIfSet(builder);
+
+        // Publishers may want to indicate that their content is child-directed and
+        // forward this information to Google.
+        Boolean isTFCD = GooglePlayServicesMediationSettings.isTaggedForChildDirectedTreatment();
+        if (isTFCD != null) {
+            builder.tagForChildDirectedTreatment(isTFCD.booleanValue());
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Fixed.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 25]
```diff
@@ -166,87 +166,91 @@ protected boolean checkAndInitializeSdk(@NonNull Activity launcherActivity,
...
-            builder.addNetworkExtrasBundle(AdMobAdapter.class, GooglePlayServicesMediationSettings.getNpaBundle());
+            builder.addNetworkExtrasBundle(AdMobAdapter.class,
+                    GooglePlayServicesMediationSettings.getNpaBundle());
         }
     }
 
     @Override
     protected boolean hasVideoAvailable() {
-        return mRewardedVideoAd != null && isAdLoaded;
+        return mRewardedAd != null && mRewardedAd.isLoaded();
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Created an `isLoaded` flag that can be used to check to avoid any threading issues.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 26]
```diff
@@ -354,27 +355,65 @@ private MoPubErrorCode getMoPubErrorCode(int error) {
...
+     *
+     * @param error Google Mobile Ads SDK Ad Request error code when showing Rewarded Video Ads.
+     * @return an equivalent MoPub SDK error code for the given Google Mobile Ads SDK Ad Request
+     * error code thrown when showing Rewarded Video Ads.
+     */
+    private MoPubErrorCode getMoPubShowErrorCode(int error) {
+        MoPubErrorCode errorCode;
+        switch (error) {
+            case RewardedAdCallback.ERROR_CODE_AD_REUSED:
+                errorCode = MoPubErrorCode.INTERNAL_ERROR;
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Fixed

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 27]
```diff
@@ -354,27 +355,65 @@ private MoPubErrorCode getMoPubErrorCode(int error) {
...
+        }
+        return errorCode;
+    }
+
     public static final class GooglePlayServicesMediationSettings implements MediationSettings {
         private static Bundle npaBundle;
         private static String contentUrl;
         private static String testDeviceId;
+        private static Boolean taggedForChildDirectedTreatment = null;
+        private static Boolean taggedForUnderAgeOfConsent = null;
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Fixed.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 28]
```diff
@@ -354,27 +355,65 @@ private MoPubErrorCode getMoPubErrorCode(int error) {
...
 
         public GooglePlayServicesMediationSettings() {
         }
 
-        public GooglePlayServicesMediationSettings(Bundle bundle) {
-            npaBundle = bundle;
-        }
+        public GooglePlayServicesMediationSettings(@NonNull Bundle bundle) {
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Updated Changelog to reflect this.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java [hunk 29]
```diff
@@ -354,27 +355,65 @@ private MoPubErrorCode getMoPubErrorCode(int error) {
...
+    }
+
     public static final class GooglePlayServicesMediationSettings implements MediationSettings {
         private static Bundle npaBundle;
         private static String contentUrl;
         private static String testDeviceId;
+        private static Boolean taggedForChildDirectedTreatment = null;
+        private static Boolean taggedForUnderAgeOfConsent = null;
 
         public GooglePlayServicesMediationSettings() {
```
[the remark this answers is not in the document]
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Removed.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java (hunk 14 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesRewardedVideo.java]: Updated.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java [hunk 30]
```diff
@@ -134,9 +157,10 @@ protected void onInvalidate() {
     private void forwardNpaIfSet(AdRequest.Builder builder) {
 
         // Only forward the "npa" bundle if it is explicitly set. Otherwise, don't attach it with the ad request.
-        if (GooglePlayServicesMediationSettings.getNpaBundle() != null &&
-                !GooglePlayServicesMediationSettings.getNpaBundle().isEmpty()) {
-            builder.addNetworkExtrasBundle(AdMobAdapter.class, GooglePlayServicesMediationSettings.getNpaBundle());
+        if (!TextUtils.isEmpty(GooglePlayServicesMediationSettings.getNpaValue())) {
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: To be consistent with the rewarded video implementation, this should not change (publishers are expected to always pass in a `Bundle` and not a `String`). The reason the rewarded video implementation was asked to be refactored was because MoPub's rewarded video does not have a working `localExtras` API (it's there but it doesn't do anything). The other ad formats have a working `localExtras` so they can just use that and minimize using the `MediationSettings`. They don't need to construct a `Bundle` when they can already take in a `Bundle` with exactly 1 <key, value> pair. We don't want the rewarded video adapter's `MediationSettings` implementation to unnecessarily scale up.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java [hunk 31]
```diff
@@ -115,9 +138,10 @@ protected void loadInterstitial(
     private void forwardNpaIfSet(AdRequest.Builder builder) {
 
         // Only forward the "npa" bundle if it is explicitly set. Otherwise, don't attach it with the ad request.
-        if (GooglePlayServicesMediationSettings.getNpaBundle() != null &&
-                !GooglePlayServicesMediationSettings.getNpaBundle().isEmpty()) {
-            builder.addNetworkExtrasBundle(AdMobAdapter.class, GooglePlayServicesMediationSettings.getNpaBundle());
+        if (!TextUtils.isEmpty(GooglePlayServicesMediationSettings.getNpaValue())) {
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Same comment as in the banner adapter.

File: AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java [hunk 32]
```diff
@@ -461,9 +493,10 @@ public void onAdFailedToLoad(int errorCode) {
         private void forwardNpaIfSet(AdRequest.Builder builder) {
 
             // Only forward the "npa" bundle if it is explicitly set. Otherwise, don't attach it with the ad request.
-            if (GooglePlayServicesMediationSettings.getNpaBundle() != null &&
-                    !GooglePlayServicesMediationSettings.getNpaBundle().isEmpty()) {
-                builder.addNetworkExtrasBundle(AdMobAdapter.class, GooglePlayServicesMediationSettings.getNpaBundle());
+            if (!TextUtils.isEmpty(GooglePlayServicesMediationSettings.getNpaValue())) {
```
Comment (chauduyphanvu1) [inline AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java]: Same comment as in the banner adapter.

Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/nativeads/GooglePlayServicesNative.java]: Reverted.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java (hunk 31 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesInterstitial.java]: Reverted.

File: AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java (hunk 30 above)
Comment (joshuajplagonera) [reply AdMob/src/main/java/com/mopub/mobileads/GooglePlayServicesBanner.java]: Reverted.
